src/schema_traverse.rs

```diff
@@ -0,0 +1,191 @@
+        if let Ok(tup) = v.downcast_exact::<PyTuple>() {
+            gather_schema(tup.get_item(0)?.downcast_exact()?, ctx)?;
+        } else {
+            gather_schema(v.downcast_exact()?, ctx)?;
+        }
+    }
+    Ok(())
+}
+
+fn gather_arguments(arguments: &Bound<'_, PyList>, ctx: &mut GatherCtx) -> PyResult<()> {
+    for v in arguments.iter() {
+        traverse_key_fn!("schema", gather_schema, v.downcast_exact::<PyDict>()?, ctx);
+    }
+    Ok(())
+}
+
+// Has 100% coverage in Pydantic side. This is exclusively used there
+#[cfg_attr(has_coverage_attribute, coverage(off))]
+fn gather_schema(schema: &Bound<'_, PyDict>, ctx: &mut GatherCtx) -> PyResult<()> {
+    let Some(type_) = get!(schema, "type") else {
+        return py_err!(PyKeyError; "Schema type missing");
+    };
+    match type_.downcast_exact::<PyString>()?.to_str()? {
+        "definition-ref" => gather_definition_ref(schema, ctx),
+        "definitions" => traverse!("schema" => gather_schema, "definitions" => gather_list; schema, ctx),
+        "list" | "set" | "frozenset" | "generator" => traverse!("items_schema" => gather_schema; schema, ctx),
+        "tuple" => traverse!("items_schema" => gather_list; schema, ctx),
+        "dict" => traverse!("keys_schema" => gather_schema, "values_schema" => gather_schema; schema, ctx),
+        "union" => traverse!("choices" => gather_union_choices; schema, ctx),
+        "tagged-union" => traverse!("choices" => gather_dict; schema, ctx),
+        "chain" => traverse!("steps" => gather_list; schema, ctx),
+        "lax-or-strict" => traverse!("lax_schema" => gather_schema, "strict_schema" => gather_schema; schema, ctx),
+        "json-or-python" => traverse!("json_schema" => gather_schema, "python_schema" => gather_schema; schema, ctx),
+        "model-fields" | "typed-dict" => traverse!(
+            "extras_schema" => gather_schema, "computed_fields" => gather_list, "fields" => gather_dict; schema, ctx
+        ),
+        "dataclass-args" => traverse!("computed_fields" => gather_list, "fields" => gather_list; schema, ctx),
+        "arguments" => traverse!(
+            "arguments_schema" => gather_arguments,
+            "var_args_schema" => gather_schema,
+            "var_kwargs_schema" => gather_schema;
+            schema, ctx
+        ),
+        "call" => traverse!("arguments_schema" => gather_schema, "return_schema" => gather_schema; schema, ctx),
+        "computed-field" | "function-plain" => traverse!("return_schema" => gather_schema; schema, ctx),
+        "function-wrap" => traverse!("return_schema" => gather_schema, "schema" => gather_schema; schema, ctx),
+        _ => traverse!("schema" => gather_schema; schema, ctx),
+    }
+}
+
+struct GatherCtx<'a, 'py> {
+    definitions: &'a Bound<'py, PyDict>,
+    meta_with_keys: Option<(Bound<'py, PyDict>, &'a Bound<'py, PySet>)>,
+    def_refs: Bound<'py, PyDict>,
+    recursive_def_refs: Bound<'py, PySet>,
+    recursively_seen_refs: HashSet<String>,
```

Comment:
There might be an optimization to store these as Python strings to avoid round-trips:

```suggestion
    recursively_seen_refs: HashSet<PyBackedStr>,
```

Reply:
Wondering does it help here as the first `contains` check anyways converts it into rust string right away?

Reply:
Changed to using PySet, it seems to be the fastest here. (Also faster than `HashSet<PyBackedStr>`)